### PR TITLE
Web Push Notifications

### LINKED ISSUE #1122: Subscribe for mobile notifications from accounts to drive engagement
A big part of what brings me back to twitter every morning are all the notifications on my phone from twitter accounts I've subscribed to. I read them like the morning paper.

If Mastodon had this, I'd be less inclined to visit the bird site and more inclined to stay just on Mastodon.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by sorin-davidoi**
This PR introduces the ability to send [Web Push Notifications](https://developers.google.com/web/fundamentals/engage-and-retain/push-notifications/web-push-protocol). The basic idea is to save a push subscription to the backend, which can then be used to wake up the Service Worker and display a notification when something happened, even if Mastodon is not opened. Together with #3052 this can make Mastodon pretty much indistinguishable from a native application (at least on Android).

Features to implement:
- [x] Register Service Worker
- [x] Generate VAPID keys in the backend
- [x] Pass the public VAPID key to the frontend
- [x] Register push subscription to the backend
- [x] Delete invalid push subscriptions
- [x] Send push notifications
  - [x] Mentions
  - [x] New followers
    - [x] Follow
    - [x] Follow request
  - [x] Favorites
  - [x] Boosts
- [x] Localize messages
- [x] Handle notification interactions
- [x] Integrate in the UI - one should be able to control the settings for each individual subscription (e.g. for what type of events it gets triggered). This is not a per account preference, as I assume people want different settings for each device (e.g. only push for mentions on mobile, but push all notifications on desktop).
- [x] Handle backend changing the `applicationServerKey`
- [x] Handle RTL
- [x] Display attachment
- [x] Unregister push subscription when logging out - can not figure out how to associate a subscription with a session, such that when the user logs out the subscription is destroyed
- [x] Set VAPID keys via configuration
- [x] Move logic out of `SettingsController` - I could not figure out how to create a new endpoint for this, so I've hijacked this controller for adding and updating subscriptions.
- [x] Add actions to the notification
  - [x] Favourite / boost status

[Browser support](http://caniuse.com/#search=push): Chrome and Firefox. Cannot be polyfilled.

Closes #1122, #1039.

### CODE REVIEW

**Comment by nolanlawson** [comment]
I was just about to ask how you managed to get Chrome to do push notifications without a GCM/FCM ID, but it looks like they've [removed that requirement](https://web-push-book.gauntface.com/chapter-06/01-non-standards-browsers/), which is going to make our lives a lot easier. Hooray! :tada:

**Comment by sorin-davidoi** [comment]
I think this is ready for a first review. Updated the description with stuff that still needs to be done. Tested on Firefox and Chrome on Linux (KDE Plasma) and on Android.

photo_2017-05-27_00-31-24
image
photo_2017-05-27_00-30-57
photo_2017-05-27_00-31-16

File: app/javascript/mastodon/web_push.js
Comment (akihikodaki) [inline app/javascript/mastodon/web_push.js]: I suppose you mean `getPushSubscription` instead of `getPushSubscription()`?

File: db/schema.rb
Comment (akihikodaki) [inline db/schema.rb]: Why not index?

File: app/controllers/api/web/push_notification_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_notification_controller.rb]: Spec?

File: app/controllers/api/web/settings_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/settings_controller.rb]: Spec this, too.

File: app/controllers/api/web/push_notification_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_notification_controller.rb]: Where `WebPushNotifications` comes from?

File: app/controllers/api/web/settings_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/settings_controller.rb]: What is it for?

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I have some questions and found something suspicious, so please review them. I'm looking forward this feature.

File: app/javascript/mastodon/web_push.js
Comment (sorin-davidoi) [inline app/javascript/mastodon/web_push.js]: Yes.

File: app/controllers/api/web/settings_controller.rb
Comment (sorin-davidoi) [inline app/controllers/api/web/settings_controller.rb]: Don't think so.

File: db/schema.rb
Comment (sorin-davidoi) [inline db/schema.rb]: This was generated by the following command, so I cannot answer that 😄  How should it look like?
```
rails generate model WebPushSubscription account_id:integer endpoint:string:unique key_p256dh:string key_auth:string
```

File: app/controllers/api/web/push_notification_controller.rb
Comment (sorin-davidoi) [inline app/controllers/api/web/push_notification_controller.rb]: Ruby magic 😄 ? Guess it is the database model?

**Comment by sorin-davidoi** [comment]
Will try to see tomorrow how Rails tests work and maybe add some specs.

File: db/schema.rb
Comment (akihikodaki) [inline db/schema.rb]: Please look at other tables. Here is an example:
```Ruby
t.index ["account_id"], name: "index_reports_on_account_id", using: :btree
```

File: app/controllers/api/web/push_notification_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_notification_controller.rb]: Ruby can sometimes do magic, but it should be in `schema.rb` anyway if it is a database model.

File: app/controllers/api/web/settings_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/settings_controller.rb]: This makes the spec fail.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I found an error with CI.

**Comment by sorin-davidoi** [comment]
If [this spec](https://www.chromestatus.com/feature/5743740178137088) gets traction we might even be able to reply from the notification.

**Comment by Gargron** [comment]
33 files change - I'm already scared :D Is this ready for review?

**Comment by sorin-davidoi** [comment]
I still haven't figured out how to attach the subscription to the session, so that when the user logs out the subscription gets deleted. Also need to pass the VAPID keys via config files / environment (right now they are generated when the server starts).

**Comment by sorin-davidoi** [comment]
@akihikodaki  I explored a bit our options for clearing the subscriptions. We have the following in `doorkeeper.rb`:
```ruby
  # Reuse access token for the same resource owner within an application (disabled by default)
  # Rationale: https://github.com/doorkeeper-gem/doorkeeper/issues/383
  reuse_access_token
```

What I understand from this is that we do not keep track of each user session (we use one token for all sessions), so there is no way to associate a push subscription with the session that created it. What is the reason for this? What would be the implications of issuing a new token for each login and deleting the token when the user logs out (attaching the push subscription to this token and deleting them together).

Another approach would be to manually delete the push subscription from the front-end when the user logs out, but that would be a huge security / privacy issue.

Don't really know how to proceed with this.

**Comment by Gargron** [comment]
>What is the reason for this? 

The web UI uses the API like any other app would use the API, so it needs an access token. Reusing of the access token was implemented in response to #1681

**Comment by akihikodaki** [comment]
> Another approach would be to manually delete the push subscription from the front-end when the user logs out, but that would create a huge security / privacy issue.

Could you tell me about your security concern?

**Comment by sorin-davidoi** [comment]
1. Call `/auth/sign_out` -> success
2. Call endpoint deleting the current subscription -> fail

Result: the user is logged out, but the browser keeps getting push notifications for that account. Also someone can mess with the code and skip the request to delete the subscription.

Signing out and deleting the subscription should be an atomic operation, handled by the back-end.

**Comment by Gargron** [comment]
http://www.jonathanleighton.com/articles/2013/revocable-sessions-with-devise/

**Comment by akihikodaki** [comment]
@sorin-davidoi 
> the user is logged out, but the browser keeps getting push notifications for that account. Also someone can mess with the code and skip the request to delete the subscription.

Even now we have `access_token`, which remains valid after signing out.

File: spec/controllers/api/web/push_subscriptions_controller_spec.rb
Comment (sorin-davidoi) [inline spec/controllers/api/web/push_subscriptions_controller_spec.rb]: The tests fail because these two lines need to be executed after each call to `sign_in`. Tried putting the following into `spec_helper.rb` but didn't help. Any pointers?

```ruby
Devise.setup do |config|
  Warden::Manager.after_set_user except: :fetch do |user, auth, opts|
    ActiveSession.deactivate warden.raw_session['auth_id']
    warden.raw_session['auth_id'] = user.activate_session
  end
end
```

**Comment by Gargron** [comment]
I suggest extracting Devise modifications into a separate PR to make reviewing easier

File: spec/controllers/api/web/push_subscriptions_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/api/web/push_subscriptions_controller_spec.rb]: I think it is because `Devise::Test::ControllerHelpers` does not support configuration by `Devise.setup`.
https://github.com/plataformatec/devise/blob/9fe7040db90be1acb499d2534b077266dbf3b209/lib/devise/test/controller_helpers.rb#L63

`Devise::Test::ControllerHelpers` is included in `rails_helper`. You may hook `sign_in` there. (it's is a bit hacky, but it cannot be helped, I think.)
https://github.com/tootsuite/mastodon/blob/fd66f7cdc0931a67a2e24d887fb79f44fc42e2bc/spec/rails_helper.rb#L26

**Review by akihikodaki — DISMISSED** [review]

**Comment by sorin-davidoi** [comment]
Will continue work once https://github.com/tootsuite/mastodon/pull/3616 is merged.

File: config/routes.rb
Comment (sorin-davidoi) [inline config/routes.rb]: @Gargron Do you have some insights into this? If I call `axios.put('/api/web/push_subscriptions/2'` I get this in the logs:

```
Started PUT "/api/web/push_subscriptions/2" for 127.0.0.1 at 2017-06-24 16:34:57 +0200
Processing by ApplicationController#raise_not_found as HTML
```

`axios.put('/api/web/push_subscriptions')` works fine though.

Cannot seem to figure it out.

Comment (Gargron) [inline config/routes.rb]: "resource" means there's no index, show with no ID (e.g. there is always one item)

"resources" is what gives you both index and show with ID

**Comment by sorin-davidoi** [comment]
Added actions when you are mentioned:

**Comment by sorin-davidoi** [comment]
I think this is ready for review. The service worker is configured **not** to cache anything. The `offline-plugin` configuration is copied from #3052 and **is only active during in the production build**. All the lint errors are logging statements that I've left in to aid testing / debugging.

**Comment by sorin-davidoi** [comment]
Any brave volunteers who would like to deploy this to a test instance? @nolanlawson , @beatrix-bitrot, heard you might fancy push notifications on mobile devices :smile:

**Comment by nightpool** [comment]
Does this only work on mobile devices? Or can I use it on desktop browsers as well?

**Comment by sorin-davidoi** [comment]
On desktop as well.

**Comment by sorin-davidoi** [comment]
On @nolanlawson's todo list for #3052:
> update tootsuite/mastodon to specify that sw.js should be served with max-age:0 (otherwise the browser may cache the sw.js for as much as 24 hours)

I suppose we would need to do this here as well. I guess this needs to go in the documentation (the example nginx configuration)? Should also be specified in the release notes I think (together with generating the VAPID keys).

**Comment by sorin-davidoi** [comment]
Unfortunately, this won't work in Samsung Internet. They support Service Worker and Push Notifications, but still rely on GCM for authentication.

**Comment by nolanlawson** [comment]
Hmm seems the only fix for Samsung Internet would be to ask admins to sign up for GCM/FCM and put their credentials in a config somewhere. The other option is to wait for Samsung to fix it, which I hope they would since no other browser requires this (Chrome used to but dropped it, and Edge won't require any such thing).

I'd vote to just wait for Samsung to fix it.

**Comment by sorin-davidoi** [comment]
Just tried the beta from the Play Store and still no support.

**Comment by nolanlawson** [comment]
Asking admins to supply their own FCM/GCM feels very un-Mastodon to me (centralized proprietary service), so maybe let's just hope Samsung fixes it.

Can we detect Samsung Internet (using something other than UA sniffing) and treat it like a browser that just doesn't support push?

**Comment by sorin-davidoi** [comment]
We don't detect it explicitly, but the setup is conservative. Will only be enabled in the UI if everything works: https://github.com/tootsuite/mastodon/pull/3243/files#diff-d5cfaccaa25810da832cba018b1257e3R113

**Comment by nightpool** [comment]
I'm not too worried about Samsung Internet—do we have substantial evidence
that people are using it? It looks like usage is about 10% that of Chrome
on Android, and half of UC browser.

A much larger holdout here is Safari for iOS, but there's not much we can
do about that seeing as they haven't even started implementing service
workers yet.

**Comment by sorin-davidoi** [comment]
Speaking of UC browser, I tried that one too. Mastodon does not seem to load at all? And there seems to be no way to debug it :man_shrugging:

**Comment by nightpool** [comment]
aaaaa I want this landed already~~~ the new popups are so nice.

**Comment by sorin-davidoi** [comment]
Blocked by #4058.

**Comment by nightpool** [comment]
@sorin-davidoi why is that blocking? isn't that just a cosmetic error with the sessions screen?

It feels pretty unrelated to this pull request for me

**Comment by sorin-davidoi** [comment]
Push subscriptions are attached to sessions (which are supposed to last until you log in). Since the sessions are created so often, you will get the "Subscription registered" multiple times a day. More important, if you enable push notifications, and a new subscriptions is created because of this problem, it will default to being turned off again.

On glitch.social, I get ~6 new sessions a day, and I have to remember to enable push subscriptions each time.

**Comment by m4sk1n** [comment]
@sorin-davidoi, this is a bit annoying, but works good…

**Comment by sorin-davidoi** [comment]
Yes, it works, but I would argue that having to enable them multiple times during the day is not a good experience. People will be confused as to why their preferences are not saved.

**Comment by sorin-davidoi** [comment]
@beatrix-bitrot I've modified this to work with #4091.

**Comment by beatrix-bitrot** [comment]
this currently needs a rebase. 

special attention should be paid to home controller and the refactoring of JS includes that was done very recently.

after that is done and CI passes I recommend this work be merged

**Comment by nightpool** [comment]
codeclimate is complaining about the console.log statements—do you think we're good to take them out?

**Comment by sorin-davidoi** [comment]
Despite all the testing, I have a feeling that this may cause problems, which will be hard to diagnose without these logs. Would it be okay to use `eslint-ignore-line` to silence them for now?

**Review by beatrix-bitrot — APPROVED** [review]
I, Beatrix Bitrot, being of sound mind and of my own volition APPROVE THESE CHANGES

**Comment by nightpool** [comment]
@sorin-davidoi it's hard to see how user-side logging is really going to help in this case, (it's hard enough to get users to *tell* us when something goes wrong, much less get them to have the dev console open at the right time) but if you think it's important to keep it, I defer to your judgement

**Comment by beatrix-bitrot** [comment]
ah, about the console log statements? i'm indifferent

for most users i don't think they be any use but on bleeding edge instances they may come in handy

maybe leave them in master for a bit and remove before the next tagged release? idk

another option is to leave them in here, then make another PR that removes them so that if anyone needs to enable them for testing all they need to do is revert the commit that removes them

idk

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Why not `params.require`?

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Use `create!`.

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (akihikodaki) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Use `update!`.

File: app/javascript/mastodon/service_worker/web_push_notifications.js
Comment (akihikodaki) [inline app/javascript/mastodon/service_worker/web_push_notifications.js]: The following statement would be enough. A promise as resolved value will be evaluated and expanded.
```JavaScript
resolve(makeRequest(event.notification, action)
  .then(() => removeActionFromNotification(event.notification, action)));
```

File: app/javascript/mastodon/service_worker/web_push_notifications.js
Comment (akihikodaki) [inline app/javascript/mastodon/service_worker/web_push_notifications.js]: Returned values does not really matter for promises.

File: app/javascript/mastodon/service_worker/web_push_notifications.js
Comment (akihikodaki) [inline app/javascript/mastodon/service_worker/web_push_notifications.js]: * Resolve or reject always.
* Do you even have to wrap in a new promise?

File: app/javascript/mastodon/web_push_subscription.js
Comment (akihikodaki) [inline app/javascript/mastodon/web_push_subscription.js]: I think you are aware of, but here I note you should be consistent, whether you use `eslint-disable-next-line` or not. If you use, apply for all other `console`. If you do not, remove this.

File: db/migrate/20170625175513_create_web_push_subscriptions.rb
Comment (akihikodaki) [inline db/migrate/20170625175513_create_web_push_subscriptions.rb]: I guess the following columns cannot not be null, can they?

File: app/services/notify_service.rb
Comment (akihikodaki) [inline app/services/notify_service.rb]: > Failing to send push notifications should not result in Internal Server Error

Yes, but that is controllers' business. If it is just to eliminate Internal Server Error, rescue in controllers.

File: spec/controllers/api/web/push_subscriptions_controller_spec.rb
Comment (akihikodaki) [inline spec/controllers/api/web/push_subscriptions_controller_spec.rb]: I think `let` is fine for these cases.

File: spec/fabricators/web_push_subscription_fabricator.rb
Comment (akihikodaki) [inline spec/fabricators/web_push_subscription_fabricator.rb]: Use Faker gem.

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I have added many comments, but they are trivial and the logic itself looks fine. I hope they will be addressed soon and the change will be merged.

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (sorin-davidoi) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Hm, does not seem to work, as this is called from an endpoint.

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (sorin-davidoi) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Same comment as `create!`.

File: app/javascript/mastodon/web_push_subscription.js
Comment (sorin-davidoi) [inline app/javascript/mastodon/web_push_subscription.js]: Will remove the other logs.

File: app/services/notify_service.rb
Comment (sorin-davidoi) [inline app/services/notify_service.rb]: Not really sure where to move the rescues then (this is called from a ton of places, all of them services - should each of them duplicate the rescues?).

File: app/controllers/api/web/push_subscriptions_controller.rb
Comment (nightpool) [inline app/controllers/api/web/push_subscriptions_controller.rb]: what's the problem with calling create from an endpoint?

Comment (sorin-davidoi) [inline app/controllers/api/web/push_subscriptions_controller.rb]: Ah, think I misunderstood the comment. I took it as to change the name of the method from `create` to `create!`. Will update it after I fix the conflicts :roll_eyes:

File: app/services/notify_service.rb
Comment (nightpool) [inline app/services/notify_service.rb]: Are we sending notifications synchronously? that seems really bad.

Comment (sorin-davidoi) [inline app/services/notify_service.rb]: To be honest I don't really know - don't have a strong grasp on the whole sidekiq thing.

File: config/webpack/production.js
Comment (akihikodaki) [inline config/webpack/production.js]: Why don't you minify?

File: spec/fabricators/web_push_subscription_fabricator.rb
Comment (akihikodaki) [inline spec/fabricators/web_push_subscription_fabricator.rb]: `Web::PushSubscription` does not have `account_id` column.

**Review by akihikodaki — DISMISSED** [review]
All problems I have requested are addressed, but I found a few other suspicious things.

**Review by Gargron — APPROVED** [review]

**Comment by sorin-davidoi** [comment]
For things that were left unaddressed here see #4200.

**Pull request merged**
